common/page.go

```diff
@@ -364,6 +363,127 @@ func (p *Page) initEvents() {
 	}()
 }
 
+func (p *Page) urlGroupingName(ctx context.Context, urlTag string) (string, bool) {
+	p.eventHandlersMu.RLock()
+
+	// If there are no handlers for EventConsoleAPICalled.
+	if _, ok := p.eventHandlers[EventPageMetricCalled]; !ok {
+		p.eventHandlersMu.RUnlock()
+
+		return "", false
+	}
+	p.eventHandlersMu.RUnlock()
+
+	var name string
+	var nameChanged bool
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	p.eventHandlersMu.RLock()
```

Comment:
Should the same lock protect the entire method instead of taking it twice here? There can be a race condition between checking the event handlers above and then looping over them below. We might not need this explanation if we go that route (I might be wrong):

```
// A handler can register another handler from within itself. This is
// the reason to unlock the mutex before calling the handler.
```

I'd use another lock for a handler if this approach I suggest is not possible.

Reply:
The reason the release the lock is called before calling the handler is because the handler can itself add another handler. This allows for that. If we didn't release the lock and a suer tried to add another handler from another handler, then it would deadlock. I've added a test to showcase that behaviour: 814f73c577ccb8e73358c873afdbf2372851ab72

Reply:
Why do we allow handlers to add handlers? Do you think it's valid use case?

Reply:
I'm not sure of what the design philosophy is for this. I've tested with PW and they allow it:

```js
// @ts-check
const { test, chromium } = require('@playwright/test');

test.describe('lifecycle', () => {
  test('lifecycle', async () => {
    const browser = await chromium.launch();
    const page = await browser.newPage();
  
    page.on('response', (response) => {
      console.log(`Received response: ${response.url()}`);
  
      page.on('request', (request) => {
        console.log(`Captured request: ${request.url()}`);
      });
    });
  
    await page.goto('https://test.k6.io');
    
    await browser.close();
  });
});
```

Reply:
👍 To streamline this and avoid mistakes, I believe we can add a helper function that checks if an event is registered because we have similar logic in [urlTagName](https://github.com/grafana/xk6-browser/blob/31050eb39c01897d346070cc4669dfdf84499183/common/page.go#L375-L382) and [onConsoleAPICalled](https://github.com/grafana/xk6-browser/blob/31050eb39c01897d346070cc4669dfdf84499183/common/page.go#L1484-L1490). By the way, in `urlTagName`, we might have forgotten to read-unlock the lock before the first `if`.

Reply:
Sure, I feel like that could be something done in the generalize issue. I've added the link to the comment 👍